### PR TITLE
Add video:double-size option

### PR DESCRIPTION
**Pull request by ssb22**

### CODE REVIEW

**Comment by scarybeasts** [comment]
Nice.
Did you think at all about doing this transform in the system display layer to potentially optimize the resize? That may not be easily feasible until the system display layers are replaced with libSDL but I thought I'd first ask your thoughts before we go with this shorter term solution.

**Comment by ssb22** [comment]
Hi, yes I did wonder about doing it at a lower level, but I couldn't find any way of getting XShmPutImage to magnify (the consensus on Stack Overflow etc seems to be that if you want XShmPutImage to magnify then you must do it yourself in the buffer).

Incidentally, if I turn on video:double-size, go into MODE 6 and do COLOUR 129:CLS (to fill the screen with white bands), I get a little flickering in the bottom border, and it's not at all obvious why.  But it's only a minor annoyance, and certainly better than having the window too small.

**Comment by scarybeasts** [comment]
Thanks. Let's land this for now (I'll review it right away). Hopefully a future move atop libSDL will give high performance (GL based?) scaling for free.

File: EXAMPLES [hunk 1]
```diff
@@ -143,6 +143,10 @@ latency of the keyboard and sound subsystems. Maybe on your system, it drives
 too hard?
 ./beebjit -0 ~/Downloads/Superior/Galaforce.ssd -opt sound:buffer=2048,bbc:wakeup-rate=100
 
+If you have a high DPI display and the BBC window is too small,
+you can try doubling the pixels (at a very small speed cost),
+./beebjit -opt video:double-size
+
```
Comment (scarybeasts) [inline EXAMPLES:149]: This seems a bit spurious as an add-on to the Galaforce example. Perhaps add a new section 14), other significant parameter examples or some such?

File: render.c [hunk 2]
```diff
@@ -108,6 +109,13 @@ render_create(struct teletext_struct* p_teletext,
   width = (640 + (border_chars * 2 * 16));
   height = (512 + (border_chars * 2 * 16));
 
+  if (util_has_option(
+      p_options->p_opt_flags, "video:double-size")) {
+    width *= 2;
+    height *= 2;
+    p_render->is_double_size = 1;
+  } else p_render->is_double_size = 0;
```
Comment (scarybeasts) [inline render.c:117]: Style: beebjit always does
if () {
} else {
}

even if the else statement is a single statement.

File: render.c [hunk 3]
```diff
@@ -656,18 +664,39 @@ render_clear_buffer(struct render_struct* p_render) {
 void
 render_double_up_lines(struct render_struct* p_render) {
   /* TODO: only need to double up partial lines within the render border. */
-  uint32_t line;
-
-  uint32_t lines = (p_render->height / 2);
   uint32_t width = p_render->width;
+  size_t line_size = width * sizeof(uint32_t);
```
Comment (scarybeasts) [inline render.c:668]: Style: parenthesis around the calculation.

File: render.c [hunk 4]
```diff
@@ -656,18 +664,39 @@ render_clear_buffer(struct render_struct* p_render) {
...
   uint32_t* p_buffer = p_render->p_buffer;
   uint32_t* p_buffer_next_line = (p_buffer + width);
-
-  for (line = 0; line < lines; ++line) {
-    (void) memcpy(p_buffer_next_line, p_buffer, (width * 4));
-    p_buffer += double_width;
-    p_buffer_next_line += double_width;
+  
+  if (p_render->is_double_size) {
+    int32_t line, column; /* must be signed */
```
Comment (scarybeasts) [inline render.c:674]: Style: one declaration per line. Comment should be init-capped, end with period.

File: render.c [hunk 5]
```diff
@@ -656,18 +664,39 @@ render_clear_buffer(struct render_struct* p_render) {
...
+    int32_t lines = p_render->height / 4;
```
Comment (scarybeasts) [inline render.c:675]: Style: parens.

File: render.c [hunk 6]
```diff
@@ -656,18 +664,39 @@ render_clear_buffer(struct render_struct* p_render) {
...
+    uint32_t half_width = width / 2;
+    for (line = 0; line < lines; ++line) {
+      for (column = half_width-1; column >= 0; --column) {
+        p_buffer[column*2] = p_buffer[column*2+1] = p_buffer[column];
```
Comment (scarybeasts) [inline render.c:679]: Style: spaces between tokens, and parens for the non-trivial calcuations, i.e.

p_buffer[column * 2] = p_buffer[(column * 2) + 1]...

File: render.c [hunk 7]
```diff
@@ -656,18 +664,39 @@ render_clear_buffer(struct render_struct* p_render) {
...
+        p_buffer[column*2] = p_buffer[column*2+1] = p_buffer[column]; 
+      }
+      (void) memcpy(p_buffer_next_line, p_buffer, line_size);
+      p_buffer += double_width;
+      p_buffer_next_line += double_width;
+    }
+    p_buffer = p_render->p_buffer;
+    lines = p_render->height / 2;
+    for (line = lines-1; line >= 0; --line) {
+      if (line) (void) memcpy(p_buffer+2*line*width, p_buffer+line*width, line_size);
```
Comment (scarybeasts) [inline render.c:688]: Style: more instances of spacing. Also, the if should be:

if () {
}

beebjit always expands like that even for simple statements.

**Comment by ssb22** [comment]
Hi, sure, I just had a go at fixing those style issues and hopefully the pull request has been updated

**Comment by scarybeasts** [comment]
Thanks again. I landed this then tweaked a few more things in https://github.com/scarybeasts/beebjit/commit/b130becb0c5ebc6830422a68fb6f84e03135ec0a

I removed an " if (line)" conditional in one of the copy loops, hopefully that didn't break anything. It didn't make sense to me unless all the lines being doubled up were treated the same.

**Comment by ssb22** [comment]
Thanks. The "if (line)" was there to avoid copying line 0 to line 2*0 which is just 0.  Copying a line to itself SHOULD have no effect, but the C standard does say memcpy() behaviour is undefined if you pass it overlapping areas, and "the exact same location" is technically overlapping, so it's not impossible somebody will come up with a library implementation that crashes when it sees this, so I was being "standards-compliant" and avoiding copying line 0 to line 0.

**Comment by scarybeasts** [comment]
Ah, good point. We do want to code correctly where memcpy()s are concerned! Shall we put it back, with a comment as to why there's that special case? (I also like avoiding the wasted work of copying a line to itself, on principle.)

**Pull request merged**